### PR TITLE
build: adjust the build to enable early syntax parser

### PR DESCRIPTION
**Pull request by compnerd**
The early syntax parser is required to support macros on Windows.  This wires up the pinned toolchain fetch, extract, and wiring to build and link against the early syntax parser.

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -401,6 +405,28 @@ function Ensure-WindowsSDK {
   }
 }
 
+function Ensure-SwiftToolchain($Arch) {
+  if (Test-Path "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe" -IsValid) {
```
Comment (tristanlabelle) [inline build.ps1:409]: Is `-IsValid` what we want? It looks like it's only a syntactic check.

image

File: build.ps1 [hunk 2]
```diff
@@ -705,6 +731,25 @@ function Build-WiXProject() {
...
+    $env:SDKROOT = "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk"
+
+    Build-CMakeProject `
+      -Src $SourceCache\swift-syntax `
+      -Bin $BinaryCache\99 `
+      -InstallTo "$($Arch.ToolchainInstallRoot)\usr" `
+      -Arch $Arch `
+      -BuildTargets default `
+      -Defines @{
+        CMAKE_Swift_COMPILER = "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe";
```
Comment (tristanlabelle) [inline build.ps1:746]: Do we need to repeat the full path if it's in the Path above?

File: build.ps1 [hunk 3]
```diff
@@ -775,6 +822,8 @@ function Build-Compilers($Arch, [switch]$Test = $false) {
         # give us a sligtly faster build.
         CMAKE_BUILD_TYPE = "Release";
         CMAKE_INSTALL_PREFIX = "$($Arch.ToolchainInstallRoot)\usr";
+        CMAKE_Swift_COMPILER = "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe";
+        CMAKE_Swift_FLAGS = "-sdk $BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk";
```
Comment (tristanlabelle) [inline build.ps1:826]: @("-sdk", "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk")

We need this so we can escape spaces properly (ie within one of the $Vars but not between -sdk and the path)

File: build.ps1 [hunk 4]
```diff
@@ -83,6 +86,7 @@ param(
   [string] $SwiftDebugFormat = "dwarf",
   [string[]] $SDKs = @("X64","X86","Arm64"),
   [string] $ProductVersion = "0.0.0",
+  [string] $PinnedToolchain = "swift-DEVELOPMENT-SNAPSHOT-2023-08-12-a",
```
Comment (tristanlabelle) [inline build.ps1:89]: maybe `PreviousToolchain`? It describes a bit better why we need it.

**Review by tristanlabelle — CHANGES_REQUESTED** [review]
The Flags part is needs to be an array.

File: build.ps1 (hunk 4 above)
Comment (compnerd) [reply build.ps1:89]: It isn't a previous toolchain, its a pinned toolchain that we use to build parts of it.

File: build.ps1 (hunk 2 above)
Comment (compnerd) [reply build.ps1:746]: Yes, we should always use absolute paths even if it is in the path.  This is super critical as we don't want accidental usage of the wrong compiler.  (`S:\b\1\bin` will always be given precedence).

File: build.ps1 (hunk 1 above)
Comment (compnerd) [reply build.ps1:409]: No, this is wrong.

File: build.ps1 (hunk 3 above)
Comment (compnerd) [reply build.ps1:826]: Ugh, good point!

File: build.ps1 (hunk 4 above)
Comment (compnerd) [reply build.ps1:89]: In some cases this may even be a future toolchain (like on 5.9).

File: build.ps1 [hunk 5]
```diff
@@ -705,6 +738,25 @@ function Build-WiXProject() {
...
+    Build-CMakeProject `
+      -Src $SourceCache\swift-syntax `
+      -Bin $BinaryCache\99 `
+      -InstallTo "$($Arch.ToolchainInstallRoot)\usr" `
+      -Arch $Arch `
+      -BuildTargets default `
+      -Defines @{
+        BUILD_SHARED_LIBS = "YES";
+        CMAKE_Swift_COMPILER = "$BinaryCache\toolchains\$PinnedToolchain\PFiles64\Swift\Toolchains\0.0.0+Asserts\usr\bin\swiftc.exe";
+        CMAKE_Swift_FLAGS = "-sdk ${env:SDKROOT}";
```
Comment (tristanlabelle) [inline build.ps1:755]: Oops, another instance of needing `@("-sdk", $env:SDKROOT)`

**Review by tristanlabelle — APPROVED** [review]
Looks good but for the Flags with spaces!

File: build.ps1 (hunk 5 above)
Comment (compnerd) [reply build.ps1:755]: I uh, yes, it must've been a clone that did this!

**Comment by hjyamauchi** [comment]
Q: does this early-build toolchain installation conflict with the toolchain that you had installed? Or is this orthogonal?

**Comment by compnerd** [comment]
It shouldn't as we explicitly take steps to download and extract the contents rather than run the installer.  This renders the toolchain inert and you must go out of your way to use it.

**Pull request merged**